### ISSUE TITLE
The version number in generated install.json is a number

### ISSUE THREAD

**Issue by chylli-deriv** [perl-carton/carton#289]
when install with following cpanfile:

```
requires 'CPAN::Audit', '== 20240117.001';
```

There is no CPAN::Audit in generated cpanfile.snapshot. That's because in the generated install.json the version number is a number. So when [json file is read as json](https://metacpan.org/release/MIYAGAWA/Carton-v1.0.35/source/lib/Carton/Snapshot.pm#L171), it will be treated as a float number, and [then the version is not matched.](https://metacpan.org/release/MIYAGAWA/Carton-v1.0.35/source/lib/Carton/Snapshot.pm#L159)

**Comment by miyagawa** [comment]
Weird. I can reproduce this with Carton and even with Carmel. However...

> That's because in the generated install.json the version number is a number. So when [json file is read as json](https://metacpan.org/release/MIYAGAWA/Carton-v1.0.35/source/lib/Carton/Snapshot.pm#L171), it will be treated as a float number,

Is it? The version looks to me a string in the `install.json` file?

```
➜  jq . ./lib/perl5/darwin-2level/.meta/CPAN-Audit-20240117.001/install.json | grep 2024
  "version": "20240117.001",
  "dist": "CPAN-Audit-20240117.001",
      "version": "20240117.001",
      "version": "20240117.001",
      "version": "20240117.001",
  "pathname": "B/BD/BDFOY/CPAN-Audit-20240117.001.tar.gz",
```

**Comment by chylli-deriv** [comment]
> Weird. I can reproduce this with Carton and even with Carmel.
> 
> > That's because in the generated install.json the version number is a number. So when [json file is read as json](https://metacpan.org/release/MIYAGAWA/Carton-v1.0.35/source/lib/Carton/Snapshot.pm#L171), it will be treated as a float number,
> 
> Is it? The version looks to me a string in the `install.json` file?
> 
> ```
> ➜  jq . ./lib/perl5/darwin-2level/.meta/CPAN-Audit-20240117.001/install.json | grep 2024
>   "version": "20240117.001",
>   "dist": "CPAN-Audit-20240117.001",
>       "version": "20240117.001",
>       "version": "20240117.001",
>       "version": "20240117.001",
>   "pathname": "B/BD/BDFOY/CPAN-Audit-20240117.001.tar.gz",
> ```

here is my result:
```
 jq . local/lib/perl5/x86_64-linux/.meta/CPAN-Audit-20240117.001/install.json  | grep version
      "version": 1.001,
      "version": 1.001
      "version": 1.001,
      "version": 1.001,
      "version": 20240117.001,
      "version": 1.001,
      "version": 1.001,
      "version": 20240117.001
      "version": 1.001
      "version": 1.001
  "version": 20240117.001,
```
when I install it directly with cpanm, it will be ok:
```
jq . local/lib/perl5/x86_64-linux/.meta/CPAN-Audit-20240117.001/install.json  | grep version
      "version": "1.001"
      "version": "1.001"
      "version": "20240117.001"
      "version": "1.001"
      "version": "1.001"
      "version": "20240117.001",
      "version": "1.001"
      "version": "1.001"
      "version": "20240117.001"
      "version": "1.001",
      "version": "1.001"
  "version": "20240117.001",
```
Maybe some depended modules are too old ?

**Comment by miyagawa** [comment]
Hmm you're right I get the unquoted numbers when I run it with Carton. 

BTW I get a similar issue with Carmel as well so I guess something might be wrong with the distribution. Looking into what's going on.

**Comment by miyagawa** [comment]
OK, the issue is the same with Carmel, the `install.json` created in the blib/meta has the version number unquoted.

```
➜  jq . ~/.carmel/5.34.1-darwin-2level/builds/CPAN-Audit-20240117.001/blib/meta/install.json | grep version
      "version": 1.001,
      "version": 1.001,
      "version": 1.001,
      "version": 20240117.001
      "version": 20240117.001,
      "version": 1.001,
      "version": 1.001,
      "version": 1.001
      "version": 1.001,
      "version": 1.001,
  "version": 20240117.001,
```

Both Carton and Carmel use Menlo::Legacy to install the modules, so that must be the difference.

**Comment by miyagawa** [comment]
```
➜  perl -MJSON::PP\ 11
JSON::PP version 11 required--this is only version 4.06.
BEGIN failed--compilation aborted.

cpanminus:
➜  grep VERSION fatlib/JSON/PP.pm
$JSON::PP::VERSION = '2.27300';
```

I reverted the version of JSON::PP to 2.27300, and now `install.json` contains versions as strings, and both Carmel and Carton work fine.

Will look into why this is broken with the new version of JSON::PP.

**Comment by miyagawa** [comment]
I confirmed that JSON::PP 2.27400 works fine, but 2.92 or later fails.

cpanm bundles 2.27300 and is not affected.

**Comment by miyagawa** [comment]
```
➜  perl -MJSON::PP -MParse::PMFile -e 'my $p = Parse::PMFile->new->parse(shift); warn encode_json({ version => $p->{"JSON::PP"}{version} })' $(perldoc -lm JSON::PP)
{"version":2.92} at -e line 1.

➜  perl -MJSON::XS -MParse::PMFile -e 'my $p = Parse::PMFile->new->parse(shift); warn encode_json({ version => $p->{"JSON::PP"}{version} })' $(perldoc -lm JSON::PP)
{"version":"2.92"} at -e line 1.
```

with JSON::PP 2.274,

```
➜  perl -MJSON::PP -MParse::PMFile -e 'my $p = Parse::PMFile->new->parse(shift); warn encode_json({ version => $p->{"JSON::PP"}{version} })' $(perldoc -lm JSON::PP)
{"version":"2.27400"} at -e line 1.
```

**Comment by miyagawa** [comment]
according to @charsbar this can be "fixed" with `PERL_JSON_PP_USE_B=1`

**Comment by miyagawa** [comment]
The core issue though is that `Parse::PMFile->parse` returns the versions in numbers (`version::->numify`), and probably Menlo/cpanm should force strings somewhere.

**Comment by chylli-deriv** [comment]
> according to @charsbar this can be "fixed" with `PERL_JSON_PP_USE_B=1`

Yes it works! 
Thanks very much.

**Comment by miyagawa** [comment]
This is fixed in Parse-PMFile-0.45, but the module is not indexed because of a PAUSE issue https://github.com/andk/pause/issues/433

so tldr, this is fixed by either:
- pinning JSON::PP to 2.27400
- upgrading Parse::PMFile to 0.45
- setting environment variable `PERL_JSON_PP_USE_B=1`.

**Comment by miyagawa** [comment]
Parse::PMFile 0.45 is now indexed.
